opm/simulators/wells/WellInterface_impl.hpp

```diff
@@ -325,6 +331,161 @@ namespace Opm
     }
 
 
+    template<typename TypeTag>
+    template<class ValueType>
+    ValueType
+    WellInterface<TypeTag>::
+    calculateBhpFromThp(const std::vector<ValueType>& rates,
+                        const int control_index,
+                        Opm::DeferredLogger& deferred_logger) const
+    {
+        // TODO: when well is under THP control, the BHP is dependent on the rates,
+        // the well rates is also dependent on the BHP, so it might need to do some iteration.
+        // However, when group control is involved, change of the rates might impacts other wells
+        // so iterations on a higher level will be required. Some investigation might be needed when
+        // we face problems under THP control.
+
+        assert(int(rates.size()) == 3); // the vfp related only supports three phases now.
+
+        const ValueType aqua = rates[Water];
+        const ValueType liquid = rates[Oil];
+        const ValueType vapour = rates[Gas];
+
+        const int vfp        = well_controls_iget_vfp(well_controls_, control_index);
+        const double& thp    = well_controls_iget_target(well_controls_, control_index);
+        const double& alq    = well_controls_iget_alq(well_controls_, control_index);
+
+        // pick the density in the top layer
+        // TODO: it is possible it should be a Evaluation
+        const double rho = perf_densities_[0];
+
+        ValueType bhp = 0. * aqua;
+        if (well_type_ == INJECTOR) {
+            const double vfp_ref_depth = vfp_properties_->getInj()->getTable(vfp)->getDatumDepth();
+
+            const double dp = wellhelpers::computeHydrostaticCorrection(ref_depth_, vfp_ref_depth, rho, gravity_);
+
+            bhp = vfp_properties_->getInj()->bhp(vfp, aqua, liquid, vapour, thp) - dp;
+         }
+         else if (well_type_ == PRODUCER) {
+             const double vfp_ref_depth = vfp_properties_->getProd()->getTable(vfp)->getDatumDepth();
+
+             const double dp = wellhelpers::computeHydrostaticCorrection(ref_depth_, vfp_ref_depth, rho, gravity_);
+
+             bhp = vfp_properties_->getProd()->bhp(vfp, aqua, liquid, vapour, thp, alq) - dp;
+         }
+         else {
+             OPM_DEFLOG_THROW(std::logic_error, "Expected INJECTOR or PRODUCER well", deferred_logger);
+         }
+
+         return bhp;
+    }
+
+
+
+    template<typename TypeTag>
+    std::vector<double>
+    WellInterface<TypeTag>::
+    computeWellPotentialWithTHP(const Simulator& ebosSimulator,
+                                const double initial_bhp, // bhp from BHP constraints
+                                const std::vector<double>& initial_potential,
+                                Opm::DeferredLogger& deferred_logger)
+    {
+        // TODO: pay attention to the situation that finally the potential is calculated based on the bhp control
+        // TODO: should we consider the bhp constraints during the iterative process?
```

Comment:
Referring to the comments above. This function should thrown a warning and return a vector of zero values.  Including the existing same function in StandardWell. The implementation is outdated.